CI: Adding build for ARM64

ossdev07 opened the pull request:
Added arm64 test support in travis-ci .
Modified environment creation by using archiconda instead of miniconda as miniconda is not supported in arm64 currently .

- [X] closes https://github.com/pandas-dev/pandas/issues/28986
- [X] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

ossdev07 committed:
added travis-ci support for arm64

Signed-off-by: ossdev07 <ossdev@puresoftware.com>

jreback commented on ci/run_tests.sh:
```diff
@@ -5,6 +5,17 @@
 # https://github.com/pytest-dev/pytest/issues/1075
 export PYTHONHASHSEED=$(python -c 'import random; print(random.randint(1, 4294967295))')
 
+if [ -n "$LOCALE_OVERRIDE" ]; then
```
this was recently removed don’t add back

jreback commented on .travis.yml:
```diff
@@ -89,6 +96,6 @@ script:
   - echo "after_script start"
-  - source activate pandas-dev && pushd /tmp && python -c "import pandas; pandas.show_versions();" && popd
-  - ci/print_skipped.py 
+  - source activate pandas-dev && pushd /tmp && if [ `uname -m` = 'aarch64' ];then VAR="python3.7" ; else VAR="python";fi && $VAR -c "import pandas; pandas.show_versions();" && popd
```
why is this needed here?

jreback commented on ci/run_tests.sh:
```diff
@@ -19,14 +30,22 @@ PYTEST_CMD="pytest -m \"$PATTERN\" -n auto --dist=loadfile -s --strict --duratio
     DISPLAY=DISPLAY=:99.0
-    PYTEST_CMD="xvfb-run -e /dev/stdout $PYTEST_CMD"
+    if [ `uname -m` = 'aarch64' ]; then
+        PYTEST_CMD="xvfb-run -e /dev/stdout pytest -m \"$PATTERN\" -s --strict --durations=10 --junitxml=test-data.xml $TEST_ARGS $COVERAGE pandas"
```
huh? this doesn’t make sense here

jreback commented on ci/run_tests.sh:
```diff
@@ -19,14 +30,22 @@ PYTEST_CMD="pytest -m \"$PATTERN\" -n auto --dist=loadfile -s --strict --duratio
     echo "uploading coverage"
-    echo "bash <(curl -s https://codecov.io/bash) -Z -c -f $COVERAGE_FNAME"
-          bash <(curl -s https://codecov.io/bash) -Z -c -f $COVERAGE_FNAME
+    echo "bash <(curl -s https://codecov.io/bash) -Z -c -F $TYPE -f $COVERAGE_FNAME"
```
why is this here?

jbrockmendel replied:
this is almost definitely a rebase mixup; this was fixed about a week ago

datapythonista replied:
please, restore the original, I think this is from a merge

jreback commented on ci/setup_env.sh:
```diff
@@ -36,9 +43,24 @@ else
+if [ `uname -m` = 'aarch64' ]; then
+   wget -q "https://github.com/Archiconda/build-tools/releases/download/0.2.3/Archiconda3-0.2.3-Linux-aarch64.sh" -O archiconda.sh
+   chmod +x archiconda.sh
+   $IS_SUDO apt-get install python-dev
```
i believe there is a travis option to run as sudo rather than do like this

jreback commented on ci/setup_env.sh:
```diff
@@ -91,8 +113,18 @@ conda list
 conda remove --all -q -y -n pandas-dev
 
 echo
+if [ `uname -m` = 'aarch64' ]; then
```
why is this needed if we r using conda from the environment file?

jreback reviewed (changes requested):
pls make minimal changes this is changing almost everything

datapythonista commented on ci/setup_env.sh:
```diff
@@ -12,8 +14,13 @@ if [[ "$(uname)" == "Linux" && -n "$LC_ALL" ]]; then
+   MINICONDA_DIR="$HOME/archiconda3"
+   IS_SUDO="sudo"
+else
+   MINICONDA_DIR="$HOME/miniconda3"
```
you can get rid of this, the directory is fine even if it's not miniconda in this case

datapythonista replied:
can you please address this comment

datapythonista commented on ci/setup_env.sh:
```diff
@@ -130,21 +169,40 @@ python setup.py build_ext -q -i -j2
+    sudo chmod -R 777 /home/travis/archiconda3/envs/pandas-dev/lib/$ARCHICONDA_PYTHON/site-packages
+    $IS_SUDO $ARCHICONDA_PYTHON -m pip install pytest-forked
+    $IS_SUDO $ARCHICONDA_PYTHON -m pip install pytest-xdist
+    $IS_SUDO $ARCHICONDA_PYTHON -m pip install --no-deps -U pip wheel setuptools
```
Whatever needs to be installed, just use the environment file

datapythonista commented on ci/setup_env.sh:
```diff
@@ -130,21 +169,40 @@ python setup.py build_ext -q -i -j2
+    sudo systemctl start mysql
+  else
+    psql -c 'create database pandas_nosetest;' -U postgres
+  fi
```
remove this please

datapythonista replied:
address this comment, and restore the original in all this part

datapythonista reviewed (changes requested):
Not sure if we should be testing ARM, since NumPy is not testing it.

But looks like some old changes got mixed with this PR. And also, I don't think we need most of the `if` statements that you have. I'd start this PR from zero, you shouldn't need much more than the new build in `.travis.yml`, the dependencies file, and the downloading the Anaconda installer for ARM.

jbrockmendel commented on ci/deps/travis-37-aarch64.yaml:
```diff
@@ -0,0 +1,22 @@
+  - python=3.7.*
+
+  # tools
+  - cython>=0.29.13
```
the error message for this build says cython is missing

ossdev07 replied:
Hi  @jbrockmendel ,
I've tried explicitly installing cython through "setup_env.sh" script. But even then it's throwing the error .
Also, i have a doubt: If i mention a dependency in .yaml file in this directory, how does the build system installs that dependency? or does it even install that dependency ? In my findings, .yaml file only checks if the mentioned dependencies are present or not .

Can you please shed light on this ?

thanks

jbrockmendel replied:
my understanding is that when you set ENV_FILE in travis.yml, that tells conda what to install.

jbrockmendel replied:
for the compilation failure, your guess is as good as mine

ossdev07 committed:
Update run_tests.sh

datapythonista commented on .travis.yml:
```diff
@@ -89,6 +96,6 @@ script:
-  - source activate pandas-dev && pushd /tmp && python -c "import pandas; pandas.show_versions();" && popd
-  - ci/print_skipped.py 
+  - source activate pandas-dev && pushd /tmp && if [ `uname -m` = 'aarch64' ];then VAR="python3.7" ; else VAR="python";fi && $VAR -c "import pandas; pandas.show_versions();" && popd
+  - ci/print_skipped.py
```
can you undo the changes to the previous two files? I guess they were tests, but in any case, we don't want this.

datapythonista commented on ci/run_tests.sh:
```diff
@@ -19,14 +19,22 @@ PYTEST_CMD="pytest -m \"$PATTERN\" -n auto --dist=loadfile -s --strict --duratio
+        PYTEST_CMD="xvfb-run -e /dev/stdout pytest -m \"$PATTERN\" -s --strict --durations=10 --junitxml=test-data.xml $TEST_ARGS $COVERAGE pandas"
+    else
+        PYTEST_CMD="xvfb-run -e /dev/stdout $PYTEST_CMD"
+    fi
```
can you undo this if please.

datapythonista commented on ci/run_tests.sh:
```diff
@@ -19,14 +19,22 @@ PYTEST_CMD="pytest -m \"$PATTERN\" -n auto --dist=loadfile -s --strict --duratio
+   sudo sh -c "$PYTEST_CMD"
+else
+    sh -c "$PYTEST_CMD"
+fi
```
undo this also

datapythonista commented on ci/setup_env.sh:
```diff
@@ -1,5 +1,7 @@
 #!/bin/bash -e
 
+IS_SUDO=""
+ARCHICONDA_PYTHON="python3.7"
```
we don't need this

datapythonista commented on ci/setup_env.sh:
```diff
@@ -36,9 +43,24 @@ else
+   echo "chmod MINICONDA_DIR"
+   $IS_SUDO chmod -R 777 $MINICONDA_DIR
+   $IS_SUDO cp $MINICONDA_DIR/bin/* /usr/bin/
+   $IS_SUDO rm /usr/bin/lsb_release
```
Just leave the wget here, and save the file as `miniconda.sh`, so you can move the chmod and the call to it outside of the if. W edon't need the installs or anything else

datapythonista commented on ci/setup_env.sh:
```diff
@@ -50,8 +72,8 @@ echo
-conda install pip  # create conda to create a historical artifact for pip & setuptools
-conda update -n base conda
+$IS_SUDO conda install pip  # create conda to create a historical artifact for pip & setuptools
+$IS_SUDO conda update -n base conda
```
please restore this

datapythonista commented on ci/setup_env.sh:
```diff
@@ -91,8 +113,18 @@ conda list
+    $IS_SUDO conda install hypothesis
+    $IS_SUDO conda install pytz
+    $IS_SUDO chmod -R 777 $MINICONDA_DIR
+fi
```
please delete this, anything ou need move it to the environment file

datapythonista commented on ci/setup_env.sh:
```diff
@@ -91,8 +113,18 @@ conda list
+
 echo "conda env create -q --file=${ENV_FILE}"
-time conda env create -q --file="${ENV_FILE}"
+time $IS_SUDO conda env create -q --file="${ENV_FILE}"
```
please restore the original

datapythonista commented on ci/setup_env.sh:
```diff
@@ -106,13 +138,17 @@ source activate pandas-dev
+    $IS_SUDO $ARCHICONDA_PYTHON -m pip uninstall -y pandas || true
+else
+    pip uninstall -y pandas || true
+fi
```
undo this please

datapythonista commented on ci/setup_env.sh:
```diff
@@ -106,13 +138,17 @@ source activate pandas-dev
 echo "remove postgres if has been installed with conda"
 echo "we use the one from the CI"
-conda remove postgresql -y --force || true
+$IS_SUDO conda remove postgresql -y --force || true
```
restore the original

datapythonista commented on ci/setup_env.sh:
```diff
@@ -121,7 +157,10 @@ conda list pandas
+if [ `uname -m` = 'aarch64' ]; then
+    sudo chmod -R 777 /home/travis/.ccache
+fi
+python setup.py build_ext -q -i
```
undo this

datapythonista commented on ci/setup_env.sh:
```diff
@@ -130,21 +169,40 @@ python setup.py build_ext -q -i -j2
+    sudo chmod -R 777 $MINICONDA_DIR
+else
+    python -m pip install --no-deps -U pip wheel setuptools
+fi
```
undo all this

datapythonista commented on ci/setup_env.sh:
```diff
@@ -130,21 +169,40 @@ python setup.py build_ext -q -i -j2
+    $IS_SUDO $ARCHICONDA_PYTHON -m pip install --no-build-isolation -e .
+else
+    python -m pip install --no-build-isolation -e .
+fi
```
and this

datapythonista reviewed (changes requested):
Can you please address all the comments and then push? Thanks!

WillAyd commented:
@ossdev07 still active? If so can you fix merge conflicts and address comments?

ossdev07 commented:
Hi @WillAyd , I am working on it. I will let you know once I am done.

ossdev07 committed:
Switched from Archiconda to Miniforge

Signed-off-by: ossdev07 <ossdev@puresoftware.com>

ossdev07 commented:
Hi @WillAyd and @datapythonista,

-  I have removed archiconda and updated it to miniforge. It does not require sudo permissions to run conda commands as it is similar to miniconda. By this we can remove all the unwanted code as suggested previously.

-  For arm64 xvfb need to be installed seperately, so added in apt-addons in travis.

-  Flags "-n auto --dist=loadfile -s --strict" mentioned in pytest command in run_tests.sh are giving timeout issue in arm64 platform. So, I have removed the same for arm64 platform.

Please review and provide feedback if any more updates are required.

datapythonista commented on ci/run_tests.sh:
```diff
@@ -20,7 +20,11 @@ if [[ $(uname) == "Linux" && -z $DISPLAY ]]; then
+if [ "${TRAVIS_CPU_ARCH}" == "arm64" ]; then
+    PYTEST_CMD="${XVFB}pytest -m \"$PATTERN\" --durations=10 --junitxml=test-data.xml $TEST_ARGS $COVERAGE pandas";
+else
+    PYTEST_CMD="${XVFB}pytest -m \"$PATTERN\" -n auto --dist=loadfile -s --strict --durations=10 --junitxml=test-data.xml $TEST_ARGS $COVERAGE pandas"
```
We surely don't want different pytest calls. Not sure why this should be different than for the rest of builds, but we've got `$TEST_ARGS` to define build specific parameters.

This file is generic, it should be untouched in this PR.

ossdev07 replied:
@datapythonista , 

Thanks for the feedback. All flags are not supported for pytest command in arm64. Command "**xvfb-run pytest -m "(not slow and not network and not clipboard)" -n auto --dist=loadfile -s --strict --durations=10 --junitxml=test-data.xml pandas**" when tested locally on arm64 platform, it is giving below error:
**ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: --strict-data-files -n --dist=loadfile pandas**

So, I have removed these flags for arm64 jobs. If we have to use $TEST_ARGS, then I suppose I have to add this in every x86_64 job or define specifically in travis file with the platform conditional clause.

Please suggest if we can use $TEST_ARGS differently.

TomAugspurger replied:
`--strict-data-files` is from pandas, in `pandas/conftest.py`. I don't see why it wouldn't be recognized by pytest, but we need to have it. Are you calling pytest from a different directory?

ossdev07 replied:
> `--strict-data-files` is from pandas, in `pandas/conftest.py`. I don't see why it wouldn't be recognized by pytest, but we need to have it. Are you calling pytest from a different directory?

@TomAugspurger , I am running pytest from same directory. I am not sure why I am getting this issue. Do I need to add any extra dependendy?

TomAugspurger replied:
I wouldn't think so.

The `-n --dist` comes from pytest-xdist. Is that installed?

datapythonista replied:
Yep, the issue is the missing dependency. To test locally, you should have a conda environment named `pandas-dev` created by running `conda env create` from the root of the repo. If you already have that, run `conda env update` to install the missing dependencies.

ossdev07 replied:
> I wouldn't think so.
> 
> The `-n --dist` comes from pytest-xdist. Is that installed?

I have verified logs and pytest-xdist is installed. It is being displayed(pytest-xdist              1.31.0                     py_0) in **conda list** . I have tested again in travis with all flags mentioned in pytest and I am facing below error:
"xvfb-run pytest -m "(not slow and not network and not clipboard)" -n auto --dist=loadfile --durations=10 --junitxml=test-data.xml pandas
============================= test session starts ==============================
platform linux -- Python 3.7.6, pytest-5.3.5, py-1.8.1, pluggy-0.12.0
rootdir: /home/travis/build/ossdev/pandas, inifile: setup.cfg, testpaths: pandas
plugins: xdist-1.31.0, forked-1.1.3, hypothesis-5.5.4
gw0 I / gw1 I / gw2 I / gw3 I / gw4 I / gw5 I / gw6 I / gw7 I / gw8 I / gw9 I / gw10 I / gw11 I / gw12 I / gw13 I / gw14 I / gw15 I / gw16 I / gw17 I / gw18 I / gw19 I / gw20 I / gw21 I / gw22 I / gw23 I / gw24 I / gw25 I / gw26 I / gw27 I / gw28 I / gw29 I / gw30 I / gw31 I
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself."

Please suggest if I am missing anything. Thanks.

datapythonista replied:
Can you merge master and push again. Looks like a problem in the azure systems, nothing to do with your changes. Relaunching the build should fix it.

ossdev07 replied:
> Can you merge master and push again. Looks like a problem in the azure systems, nothing to do with your changes. Relaunching the build should fix it.

Do I need to remove the platform check added for running pytest command and then merge into master or Do I need to add other changes you suggested without removing platform check in the run_tests.sh and push the commit? Can you please confirm.

jreback replied:
can u fix this

ossdev07 replied:
As mentioned in comment https://github.com/pandas-dev/pandas/pull/30641#discussion_r388227224, I am facing timeout issues if I added the flags and I have removed xvfb-run as I am not running any clipboard tests and need to install xvfb separately if I need to add it.

datapythonista replied:
This is incorrect, and we're not going to have different pytest calls depending on the architecture. So let's fix this first, and if there are any problems, we'll have a look and find a solution.

ossdev07 replied:
@datapythonista ,

I have added commit with suggested changes. Please review.

datapythonista commented on ci/setup_env.sh:
```diff
@@ -36,9 +36,13 @@ else
+if [ "${TRAVIS_CPU_ARCH}" == "arm64" ]; then
+  wget -q "https://github.com/conda-forge/miniforge/releases/download/4.8.2-1/Miniforge3-4.8.2-1-Linux-aarch64.sh" -O miniconda.sh;
+else
+  wget -q "https://repo.continuum.io/miniconda/Miniconda3-latest-$CONDA_OS.sh" -O miniconda.sh;
```
I'd prefer not to repeat the `wget`. You can add something like:
```bash
CONDA_URL="https://repo.continuum.io/miniconda/Miniconda3-latest-$CONDA_OS.sh"
if [ "${TRAVIS_CPU_ARCH}" == "arm64" ]; then
    CONDA_URL="https://github.com/conda-forge/miniforge/releases/download/4.8.2-1/Miniforge3-4.8.2-1-Linux-aarch64.sh"
fi

wget -q $CONDA_URL -O miniconda.sh;
```

ossdev07 replied:
@datapythonista , 

Thanks for the suggestion. I will update as suggested and add a commit.

datapythonista commented on ci/deps/travis-37-aarch64.yaml:
```diff
@@ -0,0 +1,22 @@
+name: pandas-dev
```
This file is not being used, can we remove it?

ossdev07 replied:
Yes, We can remove it. I have replaced it with travis-37-arm64.yaml.

datapythonista reviewed (commented):
@ossdev07 looks quite good now, added few more comments

jreback commented on .travis.yml:
```diff
@@ -11,7 +11,10 @@ cache:
-
+addons:
+  apt:
+    packages:
```
this has not worked welll in travis - pls remove

ossdev07 replied:
@jreback ,

Thanks for the feedback. I have added it to install xvfb. It is getting successfully installed in Travis check. May I know where can I add the apt-get command to install it.

datapythonista replied:
If we don't need it for the other builds, we shouldn't need this to this PR. Not sure why you added it, but this is required only for the clipboard tests, which you are explicitly not running in the new build.

jreback commented on ci/deps/travis-37-arm64.yaml:
```diff
@@ -0,0 +1,21 @@
+  - pytest-xdist>=1.21
+  - hypothesis>=3.58.0
+
+  # pandas dependencies
```
are there other pandas optional deps that are available?

jreback commented on ci/setup_env.sh:
```diff
@@ -36,9 +36,13 @@ else
+fi
 chmod +x miniconda.sh
-./miniconda.sh -b
+./miniconda.sh -b -p $MINICONDA_DIR
```
why this change? u should already be in the correct dir

ossdev07 replied:
In the case of the arm64 platform, I am using miniforge3 to install conda. When we run miniconda.sh(Miniforge3-4.8.2-1-Linux-aarch64.sh) for arm64 platform, it will get installed in $HOME/miniforge3 folder whereas for x64 platform it will get installed in $HOME/miniconda3. So, I am providing the installation path as $HOME/miniconda3.

datapythonista replied:
May be you can you add a short comment before this line, so this information is explicit in the code.

datapythonista reviewed (commented):
@ossdev07 can you address the pending comments please, this is close now.

ossdev07 commented:
@datapythonista , 

Sorry, for the extra commits that have been added. Rebase has gone wrong and all commits got added. The actual commit which needs to be committed is https://github.com/pandas-dev/pandas/pull/30641/commits/887591e8a5a529ef250d3a67364280008b0e14c5.

datapythonista commented:
> Sorry, for the extra commits that have been added. Rebase has gone wrong and all commits got added. The actual commit which needs to be committed is [887591e](https://github.com/pandas-dev/pandas/commit/887591e8a5a529ef250d3a67364280008b0e14c5).

You'll have to fix your branch for us to merge this. I wrote some time ago on what I usually do in these cases in this post: https://datapythonista.me/blog/useful-git-commands.html

Hope that helps.

ossdev07 commented:
> > Sorry, for the extra commits that have been added. Rebase has gone wrong and all commits got added. The actual commit which needs to be committed is [887591e](https://github.com/pandas-dev/pandas/commit/887591e8a5a529ef250d3a67364280008b0e14c5).
> 
> You'll have to fix your branch for us to merge this. I wrote some time ago on what I usually do in these cases in this post: https://datapythonista.me/blog/useful-git-commands.html
> 
> Hope that helps.

Thanks @datapythonista. I have resolved the conflicts in the commits and added the commit with changes suggested. Please review and provide feedback if anything needs to be updated.

datapythonista commented on .travis.yml:
```diff
@@ -11,7 +11,6 @@ cache:
   directories:
     - $HOME/.cache # cython cache
     - $HOME/.ccache # compiler cache
-
```
Can you restore this deleted line please.

ossdev07 committed:
Switched from Archiconda to Miniforge

datapythonista commented on ci/setup_env.sh:
```diff
@@ -36,9 +36,17 @@ else
 
-wget -q "https://repo.continuum.io/miniconda/Miniconda3-latest-$CONDA_OS.sh" -O miniconda.sh
+if [ "${TRAVIS_CPU_ARCH}" == "arm64" ]; then
+  apt-get update && apt-get install xvfb;
```
Thanks for the update @ossdev07, much better now.

I see now. Apparently, for the rest of the architectures `xvfb` is installed by default, but not for `aarch64`. I thought Travis would be more consistent, but I see your point now.

I know we asked you to remove it, but can you please restore this in the yaml as `services: xvfb` please? And remove this line.

Also, can you remove the semicolon you added at the end of the lines in this file `CONDA_URL=...`, and `wget ...`.

ossdev07 replied:
@datapythonista ,

Thanks for the feedback. I will update as suggested.

pruthviteja replied:
@datapythonista ,

I have updated the commit with changes suggested. Please review.

ossdev07 replied:
@datapythonista ,

I have updated the commit with changes suggested. Please review.

rhenwood-arm commented:
(I work with @ossdev07 ) : @ossdev07 as an independent task, can you make a note of the missing `xvfb` for Travis-CI in their community forums: https://travis-ci.community/

ossdev07 commented:
> (I work with @ossdev07 ) : @ossdev07 as an independent task, can you make a note of the missing `xvfb` for Travis-CI in their community forums: https://travis-ci.community/

@rhenwood-arm ,

I will take a note of it and will raise it in the travis community.

datapythonista commented:
Travis is red, can you have a look and see what's wrong please

ossdev07 commented:
> Travis is red, can you have a look and see what's wrong please

@datapythonista ,

After adding all flags in pytest command, it is failing with the timeout issue at different steps for each build for arm64 platform. That is the reason I have removed those three flags.

Could you please suggest if there is any resolution to run without these flags for a specific platform?

datapythonista commented:
IIRC the flags you changed were for `pytest`, and in this case `pytest` didn't even started, so that's surely not going to solve the problem.

The build was stalled during `ci/submit_cython_cache.sh`. Can you merge master, so we can see if the problem happens at the same stage. I'm wondering if Travis is reliable for ARM64...

ossdev07 commented:
> IIRC the flags you changed were for `pytest`, and in this case `pytest` didn't even started, so that's surely not going to solve the problem.
> 
> The build was stalled during `ci/submit_cython_cache.sh`. Can you merge master, so we can see if the problem happens at the same stage. I'm wondering if Travis is reliable for ARM64...

@datapythonista , 

After merging master build is stalled during `ci/submit_cython_cache.sh` in travis check in this PR, whereas in personal travis, it is getting stalled during `pytest`. Please check report https://travis-ci.org/github/ossdev07/pandas/jobs/664267609 for the same. 

Can you please suggest if I am missing anything? Thanks.

datapythonista commented:
Seems like a problem in Travis to me.

May be you can comment the two lines about the cache in `travis.yml`, and see if that works:
```
  # - ci/prep_cython_cache.sh
  - ci/setup_env.sh
  # - ci/submit_cython_cache.sh
```
May be that helps us understand better the problem. But I think you'll have to report it to Travis, doesn't seem to be a problem in our side.

ossdev07 commented:
> Seems like a problem in Travis to me.
> 
> May be you can comment the two lines about the cache in `travis.yml`, and see if that works:
> 
> ```
>   # - ci/prep_cython_cache.sh
>   - ci/setup_env.sh
>   # - ci/submit_cython_cache.sh
> ```
> 
> May be that helps us understand better the problem. But I think you'll have to report it to Travis, doesn't seem to be a problem in our side.

I have commented cache part and checked in local travis. It is getting timed out during pytest phase. 
Report link: https://travis-ci.org/github/ossdev07/pandas/jobs/664711811.

datapythonista commented:
Can you report the problem to TravisCI? I don't see what we can do in our side if their systems stall during the build in a somehow random way.

ossdev07 commented:
> Can you report the problem to TravisCI? I don't see what we can do in our side if their systems stall during the build in a somehow random way.

Thanks for the suggestion. I will raise an issue in travis community regarding the same and update if anything is required.

ossdev07 commented:
Hi @datapythonista ,

I have raised a ticket https://travis-ci.community/t/travis-build-gets-stalled-for-arm64-job-in-pandas-package/7678 in travis community for the build issue. 

As suggested in that ticket, I have rebased and added travis_wait to increase the timeout before each script. But I was facing pandas compilation issue.

**Error:**
`pandas/_libs/src/ujson/lib/ultrajsonenc.c:726:59: error: ‘fastcall’ attribute directive ignored [-Werror=attributes]
                                                           char *end) {
                                                           ^
cc1: all warnings being treated as errors
gcc -pthread -shared -B /home/travis/miniconda3/envs/pandas-dev/compiler_compat -L/home/travis/miniconda3/envs/pandas-dev/lib -Wl,-rpath=/home/travis/miniconda3/envs/pandas-dev/lib -Wl,--no-as-needed -Wl,--sysroot=/ build/temp.linux-aarch64-3.7/pandas/io/sas/sas.o -o /home/travis/build/ossdev07/pandas/pandas/io/sas/_sas.cpython-37m-aarch64-linux-gnu.so`

I have tried after removing travis_wait as well but still facing the same issue. Please check the report https://travis-ci.org/github/ossdev07/pandas/jobs/671601371 for the same.

Can you suggest any pointers on resolving this issue?

datapythonista commented:
No idea to be honest, but looks like an issue specific to ARM64.

I don't think the `travis_wait` will fix the problems, but let's try.

ossdev07 commented:
@datapythonista ,

The compilation issue is resolved when we remove "-Werror" flag in extra_compile_args variable in setup.py. But added travis_wait for running tests, but still facing the same timeout issue. If we remove the previously mentioned flags in pytest command, it is working successfully. Can you please suggest what I may be missing?

Report: https://travis-ci.org/github/ossdev07/pandas/jobs/674356951

datapythonista commented:
If it works without `-Werror` it means that the error is a warning. But we don't allow warnings in the build anyway.

But I think the problem here is the timeout. Not sure if it's a problem with Travis ARM64 machines that they stall, or if they are just too slow to compile, run the tests...

ossdev07 commented:
@datapythonista ,

I have tried running locally again and tests are triggered successfully:

**xvfb-run pytest -m "(not slow and not network and not clipboard)" -n auto --dist=loadfile -s --strict --durations=10 --junitxml=test-data.xml pandas**
===================== test session starts ==============================
platform linux -- Python 3.8.2, pytest-5.4.1, py-1.8.1, pluggy-0.12.0
rootdir: /pruthvi/pandas, inifile: setup.cfg, testpaths: pandas
plugins: hypothesis-5.8.3, forked-1.1.3, xdist-1.31.0
gw0 [68456] / gw1 [68456] / gw2 [68456] / gw3 [68456]
=4 failed, 63509 passed, 3907 skipped, 1044 xfailed, 7 xpassed, 29 warnings in 1357.02s (0:22:37) =

I think, in travis issue is with xvfb. It is not getting started.
**sudo systemctl start xvfb
Failed to start xvfb.service: Unit xvfb.service not found.**

In travis test run is triggered without xvfb-run:
**pytest -m "(not slow and not network and not clipboard)" -n auto --dist=loadfile -s --strict --durations=10 --junitxml=test-data.xml pandas**
=== test session starts ====
platform linux -- Python 3.7.6, pytest-5.4.1, py-1.8.1, pluggy-0.12.0
rootdir: /home/travis/build/ossdev07/pandas, inifile: setup.cfg, testpaths: pandas
plugins: hypothesis-5.8.3, forked-1.1.3, xdist-1.31.0
gw0 I / gw1 I / gw2 I / gw3 I / gw4 I / gw5 I / gw6 I / gw7 I / gw8 I / gw9 I / gw10 I / gw11 I / gw12 I / gw13 I / gw14 I / gw15 I / gw16 I / gw17 I / gw18 I / gw19 I / gw20 I / gw21 I / gw22 I / gw23 I / gw24 I / gw25 I / gw26 I / gw27 I / gw28 I / gw29 I / gw30 I / gw31 I

I suppose it is getting timeout due to this reason. I have added xvfb as services but still getting issue in starting xvfb services.

Can you suggest on the same?

datapythonista commented:
I don't think `xvfb` is the problem, but I may be wrong. Let me explain about `xvfb`.

pandas is able to get and set data from and to the system clipboard. In Windows that is done "directly", but Linux doesn't have a clipboard itself, but it's the X server that usually provides it. In a desktop machine, if you're using pandas locally with your browser, you'll have an X server running. But in the CI (or when using pandas in a server), the Linux doesn't have more than the terminal, so no X server exists. `xvfb` is a "simulator" of the X server. So, when we call the tests, we wrap the call with `xvfb-run`, so the clipboard tests have an X server to use, and tests don't fail.

No idea on whether arm64 uses X server, or what. But you can control when we use `xvfb` with this condition: https://github.com/pandas-dev/pandas/blob/master/ci/run_tests.sh#L18

Also, we're able to control whether we run the clipboard tests in a build using `not clipboard`, like in here: https://github.com/pandas-dev/pandas/blob/master/ci/azure/posix.yml#L21

So, if what you need, is to not use `xvfb`, that's fine, and you can skip the clipboard tests in arm if they don't work. But I think the problem is more that the Travis hardware is just too slow, and it's cancelled after some step takes more than 10 minutes without returning any output.

ossdev07 commented:
@datapythonista ,

I have added travis_wait and installed xvfb. Tests are getting exited with below error:

xvfb-run pytest -m "(not slow and not network and not clipboard)" -n auto --dist=loadfile -s --strict --durations=10 --junitxml=test-data.xml pandas
===== test session starts ================
platform linux -- Python 3.7.6, pytest-5.4.1, py-1.8.1, pluggy-0.12.0
rootdir: /home/travis/build/ossdev07/pandas, inifile: setup.cfg, testpaths: pandas
plugins: hypothesis-5.8.3, forked-1.1.3, xdist-1.31.0
gw0 I / gw1 I / gw2 I / gw3 I / gw4 I / gw5 I / gw6 I / gw7 I / gw8 I / gw9 I / gw10 I / gw11 I / gw12 I / gw13 I / gw14 I / gw15 I / gw16 I / gw17 I / gw18 I / gw19 I / gw20 I / gw21 I / gw22 I / gw23 I / gw24 I / gw25 I / gw26 I / gw27 I / gw28 I / gw29 I / gw30 I / gw31 I
[gw14] node down: Not properly terminated
replacing crashed worker gw14
[gw1] node down: Not properly terminated
replacing crashed worker gw1
[gw7] node down: Not properly terminated
replacing crashed worker gw7
[gw11] node down: Not properly terminated
replacing crashed worker gw11
[gw6] node down: Not properly terminated
replacing crashed worker gw6
[gw13] node down: Not properly terminated
replacing crashed worker gw13
The command "travis_wait 30 ci/run_tests.sh" exited with 137.

**Report:** https://travis-ci.org/github/ossdev07/pandas/jobs/683298787

I suppose it is giving error with "-n auto" flag. Could you provide me any suggestions on the above issue.

datapythonista commented:
Thanks @ossdev07 

For what I've seen, there is a discussion now on whether we want to support new architectures or not in #33971. You may want to have a look there, not sure if it's worth continuing with this until a decision is made there.

For the error, seems again a problem with Travis infrastructure. Not sure why the testing process is crashing, but I guess it's the OS killing it because of lack of resources, or something like that. Not sure if removing the `-n auto` will solve the problem, but can provide more information (the log is not so verbose when tests are running in parallel).

TomAugspurger commented:
FWIW, I don't think https://github.com/pandas-dev/pandas/issues/33971 necessarily should hold this PR up. I think whatever rules we establish, ARM64 will qualify for inclusion.

ossdev07 commented:
@datapythonista ,

> For the error, seems again a problem with Travis infrastructure. Not sure why the testing process is crashing, but I guess it's the OS killing it because of lack of resources, or something like that. Not sure if removing the `-n auto` will solve the problem, but can provide more information (the log is not so verbose when tests are running in parallel).

I have verified with parallel flag as "-n 4" instead of auto and without travis_wait, tests are triggering and passing in both amd64 and arm64 jobs.
I have just modified at line https://github.com/ossdev07/pandas/blob/arm64_test_sample/ci/run_tests.sh#L23

**Travis Report:** https://travis-ci.org/github/ossdev07/pandas/jobs/683704969

Can we proceed with these changes or flag must be auto only?

datapythonista commented:
That's interesting. I don't see those changes in this PR. Did you revert them, or tried somewhere else? I'm in my phone, so may be I'm just not seeing them.

The auto means to use and many workers as cores.

If you don't mind, there are two things I'd like to do:

We can add in the yaml a variable for the number of workers (e. g. `PYTEST_WORKERS`). Just set it to `4` for the arm build. Then in the test script, add an `if`, and set it to `auto`, if the variable is not set. This should make the arm build pass, without changing the workers in the other builds.

In parallel, would be good to find out how many workers `auto` means in arm. I guess it can be more than 8, and the workers crash because of memory (the more the workers, the more memory used, and if we run out of memory, the workers will crash). If you can also check the amount of memory in arm, and in Linux for comparison, that would be great. Understanding this will be very helpful. We want to optimise the number of workers, so the build doesn't take too long. This is taking 45 minutes, and it's by far the slowest build. Which means that for any PR, we'll have to wait for the arm build before being able to merge, if we merge the way you propose. And also, we want to understand the memory, we don't want to merge this being in the edge, and that the next PR adds a test, and this starts failing because of lack of memory.

Thanks for the work on this, and sorry it's taking so long.

ossdev07 commented:
@datapythonista ,

> We can add in the yaml a variable for the number of workers (e. g. `PYTEST_WORKERS`). Just set it to `4` for the arm build. Then in the test script, add an `if`, and set it to `auto`, if the variable is not set. This should make the arm build pass, without changing the workers in the other builds.
> 

- I have added PYTEST_WORKERS as a global variable with value "auto" and reassigning it in arm64 job. I have tested with -n value as 12 and 16 as well but workers are getting crashed and it is working with 8 workers.
- I have added xvfb installation separately only for arm64 as it is not getting installed as services. 
- Flag "-Werror" had to be removed as compilation of external dependencies were failing with warnings treated as error. 

Please check the add commit and provide feedback if anything needs to be changed.

datapythonista commented on setup.py:
```diff
@@ -433,7 +433,7 @@ def run(self):
         extra_link_args.append("/DEBUG")
 else:
-    extra_compile_args = ["-Werror"]
+    extra_compile_args = []
```
We'll have to restore this before merging, but let's have a look and see how the arm build goes for now.

ossdev07 replied:
I am confused that travis build has failed here whereas it has passed in https://travis-ci.org/github/ossdev07/pandas/builds/683838496 which has been triggered by the same commit.

Flag "-Werror" gives a compilation error while compiling external dependencies in arm64 job.

ossdev07 replied:
I have added a commit by restoring "-Werror" flag, it is giving an error in pandas compilation.

**Error:**
pandas/_libs/src/ujson/lib/ultrajsonenc.c:397:1: error: ‘fastcall’ attribute directive ignored [-Werror=attributes]
 Buffer_AppendShortHexUnchecked(char *outputOffset, unsigned short value) {

I have checked pandas/_libs/src/ujson/lib/ultrajson.h file for attribute "FASTCALL_ATTR" definition which is throwing a warning and which returns as error due to -Werror flag. Could you suggest anything on resolving this issue?

datapythonista replied:
I don't know much about of that part of the code, sorry. But it must be something specific to arm, since it's working well with all other architectures.

WillAyd replied:
The macro definition for that is going to be in ultrajson.h . Might be some other platform check needed there

WillAyd replied:
My guess is has something to do with the definition of `__x86_64__` on the platform

ossdev07 replied:
Thanks @WillAyd for the suggestion. I have added the definition as "if !defined __x86_64__ && !defined __aarch64__" in pandas/_libs/src/ujson/lib/ultrajson.h.

ossdev07 replied:
@datapythonista / @WillAyd ,
Arm build is passing but 2 tests are failing in 2 amd64 jobs. But I think they are failing for network-specific reasons rather than adding arm jobs. 

Can we merge current changes or any more updates are required?

jreback commented on .travis.yml:
```diff
@@ -33,6 +35,10 @@ matrix:
 
+    - arch: arm64
+      env:
+        - JOB="3.7" PYTEST_WORKERS=8 ENV_FILE="ci/deps/travis-37-arm64.yaml" PATTERN="(not slow and not network and not clipboard)"
```
you are likely to have much better performance with 2 workers here. I would guess that's what we have physically, and what we use on other processors.

ossdev07 replied:
I have updated to 2 workers as suggested. Please review.

jreback replied:
it still using 8 on the arm build - make this 2

and taking 56 minutes which is too long 

put the build in the allowed failures section

ossdev07 replied:
I have updated to 2 workers in the previous commit, but the build was getting stalled and failing while running tests. So, I have moved back to 8 workers. Pandas compilation and test execution are taking more time to complete in ARM64 when compared to the AMD64 platform.

jreback reviewed (changes requested):
you need to merge master, lots of things are failing.

jreback commented:
these changes are breaking other builds: https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=35039&view=logs&j=bef1c175-2c1b-51ae-044a-2437c76fc339&t=770e7bb1-09f5-5ebf-b63b-578d2906aac9&l=17

cphang99 commented:
Do the workarounds described in https://github.com/pandas-dev/pandas/issues/17792 need to be incorporated in order for the test suite to pass?

cphang99 commented:
It looks like it does. many of the failing tests appear to be `DateTime` tests as described in #17792

More details in https://gitlab.com/libreml/libreml/-/issues/110#note_339463783

ossdev07 commented:
> these changes are breaking other builds: https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=35039&view=logs&j=bef1c175-2c1b-51ae-044a-2437c76fc339&t=770e7bb1-09f5-5ebf-b63b-578d2906aac9&l=17

I have added global variable PYTEST_WORKERS in azure pipelines yml file and issues have been resolved with my azure account. Please review the changes and let me know if any more changes required.

ossdev07 commented:
@datapythonista / @jreback / @WillAyd ,

All checks are passing as well now. Please review and let me know if anything needs to be updated.

jreback commented:
this build takes 56 minutes which is way longer than the longest build

jreback commented on .travis.yml:
```diff
@@ -62,9 +61,12 @@ matrix:
-  - dist: bionic
-    python: 3.9-dev
-    env:
+    - arch: arm64
```
I think you have have to have it in *both* sections (allow_failures) and the top section; they must be identical.

ossdev07 replied:
Thanks for the suggestion. I have added it as suggested and all checks are passing as well. Please review and let me know if any more updates are required.

jreback reviewed (approved):
lgtm. @TomAugspurger ok here

TomAugspurger commented:
Thanks. So to confirm: we're putting it under allowed failures just because it takes longer? We'd consider any failures on that job to be a blocker to merging / cause for reverting if a failure is noticed?

@ossdev07 will you be available to respond to build / CI issues as they come up?

ossdev07 commented:
> Thanks. So to confirm: we're putting it under allowed failures just because it takes longer? We'd consider any failures on that job to be a blocker to merging / cause for reverting if a failure is noticed?
> 
> @ossdev07 will you be available to respond to build / CI issues as they come up?

Yes, I will be available to respond if any issues come up.

jreback commented:
yeah i think under allowed failures for now until we can get the time to be more reasonable see #34115 

@ossdev07 feel free to comment on the issue about building wheels for this (note we would have to actualy run this on the azure side and not here, is this possible?)

can you create an issue about moving this out of allowed failures (we can do this as soon as we see this working reasonbaly for a little)

jreback committed:
Merge branch 'master' into pandas_ci_arm64

ossdev07 commented:
> yeah i think under allowed failures for now until we can get the time to be more reasonable see #34115
> 
> @ossdev07 feel free to comment on the issue about building wheels for this (note we would have to actualy run this on the azure side and not here, is this possible?)
> 
> can you create an issue about moving this out of allowed failures (we can do this as soon as we see this working reasonbaly for a little)

Thanks for the approval. I will raise a ticket about moving this out of allowed failures.
